File: java/src/main/java/benchmarks/Base64Decode.java
package benchmarks;

import java.util.Base64;
import java.nio.charset.StandardCharsets;

public class Base64Decode extends Benchmark {
    private int n;
    private String str2;
    private byte[] bytes;
    private long resultVal;

    public Base64Decode() {
        n = (int) configVal("size");
        resultVal = 0L;
    }

    @Override
    public String name() {
        return "Base64::Decode";
    }

    @Override
    public void prepare() {
        String str = "a".repeat(n);
        str2 = Base64.getEncoder().encodeToString(str.getBytes());
        bytes = Base64.getDecoder().decode(str2);
    }

    @Override
    public void run(int iterationId) {
        bytes = Base64.getDecoder().decode(str2);
        resultVal = (resultVal + bytes.length) & 0xFFFFFFFFL;
    }

    @Override
    public long checksum() {
        String str3 = new String(bytes, 0, 5, StandardCharsets.UTF_8);
        String prefix2 = str2.length() > 4 ? str2.substring(0, 4) + "..." : str2;
        String prefix3 = str3.length() > 4 ? str3.substring(0, 4) + "..." : str3;
        String message = "decode " + prefix2 + " to " + prefix3 + ": " + resultVal;
        return Helper.checksum(message);
    }
}

File: nim/src/benchmarks/base64decode.nim
import std/[base64, strutils]
import ../benchmark
import ../helper

type
  Base64Decode* = ref object of Benchmark
    str2: string
    str3: string
    resultVal: uint32

proc newBase64Decode(): Benchmark =
  Base64Decode()

method name(self: Base64Decode): string = "Base64Decode"

method prepare(self: Base64Decode) =
  let n = self.config_val("size")
  let str = repeat('a', n.int)

  self.str2 = encode(str)

  try:
    self.str3 = decode(self.str2)
  except Exception:
    self.str3 = ""

  self.resultVal = 0

method run(self: Base64Decode, iteration_id: int) =
  try:
    self.str3 = decode(self.str2)
    self.resultVal = self.resultVal + uint32(self.str3.len)
  except Exception:
    discard

method checksum(self: Base64Decode): uint32 =
  var ss = "decode "
  if self.str2.len > 4:
    ss.add(self.str2[0..3] & "...")
  else:
    ss.add(self.str2)

  ss.add(" to ")

  if self.str3.len > 4:
    ss.add(self.str3[0..3] & "...")
  else:
    ss.add(self.str3)

  ss.add(": " & $self.resultVal)
  checksum(ss)

registerBenchmark("Base64Decode", newBase64Decode)
